src/NHibernate/Util/AsyncReaderWriterLock.cs

```diff
@@ -0,0 +1,219 @@
+			}
+
+			if (waitForReadLocks)
+			{
+				_readLockSemaphore.Wait();
+			}
+
+			return _writerReleaser;
+		}
+
+		public async Task<IDisposable> WriteLockAsync()
+		{
+			if (!CanEnterWriteLock(out var waitForReadLocks))
+			{
+				await _writeLockSemaphore.WaitAsync().ConfigureAwait(false);
+				lock (_writeLockSemaphore)
+				{
+					_writersWaiting--;
+				}
+			}
+
+			if (waitForReadLocks)
+			{
+				await _readLockSemaphore.WaitAsync().ConfigureAwait(false);
+			}
+
+			return _writerReleaser;
+		}
+
+		public IDisposable ReadLock()
+		{
+			if (CanEnterReadLock())
+			{
+				return _readerReleaser;
+			}
+
+			_waitingReadLockSemaphore.Wait();
+			ReleaseWaitingReader();
+
+			return _readerReleaser;
+		}
+
+		public Task<IDisposable> ReadLockAsync()
+		{
+			return CanEnterReadLock() ? _readerReleaserTask : ReadLockInternalAsync();
+
+			async Task<IDisposable> ReadLockInternalAsync()
+			{
+				await _waitingReadLockSemaphore.WaitAsync().ConfigureAwait(false);
+				ReleaseWaitingReader();
+
+				return _readerReleaser;
+			}
+		}
+
+		private bool CanEnterWriteLock(out bool waitForReadLocks)
+		{
+			waitForReadLocks = false;
+			lock (_writeLockSemaphore)
+			{
+				if (_writeLockSemaphore.CurrentCount > 0 && _writeLockSemaphore.Wait(0))
+				{
+					waitForReadLocks = _currentReaders > 0;
+					return true;
+				}
+
+				_writersWaiting++;
+			}
+
+			return false;
+		}
+
+		private void ExitWriteLock()
+		{
+			lock (_writeLockSemaphore)
+			{
+				if (_writeLockSemaphore.CurrentCount == 1)
+				{
+					throw new InvalidOperationException();
+				}
+
+				// Writers have the highest priority even if they came last
+				if (_writersWaiting > 0 || _waitingReadLockSemaphore == null)
+				{
+					_writeLockSemaphore.Release();
+					return;
+				}
+
+				if (_readersWaiting > 0)
+				{
+					_currentReaders += _readersWaiting;
+					_waitingReadLockSemaphore.Release(_readersWaiting);
+				}
+
+				_writeLockSemaphore.Release();
+			}
+		}
+
+		private bool CanEnterReadLock()
+		{
+			lock (_writeLockSemaphore)
+			{
+				if (_writersWaiting == 0 && _writeLockSemaphore.CurrentCount > 0)
+				{
+					_currentReaders++;
+
+					return true;
+				}
+
+				if (_waitingReadLockSemaphore == null)
+				{
+					_waitingReadLockSemaphore = new SemaphoreSlim(0);
+				}
+
+				_readersWaiting++;
+
+				return false;
+			}
+		}
+
+		private void ExitReadLock()
+		{
+			lock (_writeLockSemaphore)
+			{
+				if (_currentReaders == 0)
+				{
+					throw new InvalidOperationException();
+				}
+
+				_currentReaders--;
+				if (_currentReaders == 0 && _writeLockSemaphore.CurrentCount == 0)
+				{
+					_readLockSemaphore.Release();
+				}
+			}
+		}
+
+		private void ReleaseWaitingReader()
+		{
+			lock (_writeLockSemaphore)
+			{
+				_readersWaiting--;
+				if (_readersWaiting != 0)
+				{
+					return;
+				}
+
+				_waitingReadLockSemaphore.Dispose();
+				_waitingReadLockSemaphore = null;
+			}
+		}
+
+		private struct Releaser : IDisposable
```

Comment:
There is no point making this struct if it is a private. It will be boxed and leave in the heap.

Reply:
You are right, I changed it to a sealed class.

Reply:
I did some perf testing. The fastest (and free of additional allocations) method is to return the struct. The second place is class (does not matter if it is returned as class or as `IDisposable`) and third place is a struct boxed to `IDisposable`.

Reply:
Updated the code to use the fastest option. We could modify the struct to be `readonly`, but it requires C# langauge to be upgraded to `7.2` and I am not sure if it would perform any better. Also, I didn't apply the change for `AsyncLock`, as it would be a breaking change.